File: solutions/scala/120-lowest-common-ancestor-of-a-binary-search-tree.scala
/**
 * Problem 120: Lowest Common Ancestor of a Binary Search Tree (LeetCode 235)
 * Difficulty: Med
 * Language: Scala
 */
def lcaBST(root: Tree[Int], p: Int, q: Int): Option[Int] = root match {
  case Leaf => None
  case Node(l, v, _) if p < v && q < v => lcaBST(l, p, q)
  case Node(_, v, r) if p > v && q > v => lcaBST(r, p, q)
  case Node(_, v, _) => Some(v)
}

File: solutions/rust/120-lowest-common-ancestor-of-a-binary-search-tree.rs
/**
 * Problem 120: Lowest Common Ancestor of a Binary Search Tree (LeetCode 235)
 * Difficulty: Med
 * Language: Rust
 */
fn lca_bst(root: &Option<Box<TreeNode>>, p: i32, q: i32) -> Option<i32> {
    match root {
        None => None,
        Some(node) => {
            if p < node.val && q < node.val { lca_bst(&node.left, p, q) }
            else if p > node.val && q > node.val { lca_bst(&node.right, p, q) }
            else { Some(node.val) }
        }
    }
}
